Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,647 @@
+# Strongly Typed Regex Captures
+
+<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
+**Table of Contents**
+
+- [Strongly typed regex captures](#strongly-typed-regex-captures)
+    - [Introduction](#introduction)
+    - [Motivation](#motivation)
+    - [Proposed solution](#proposed-solution)
+    - [Detailed design](#detailed-design)
+        - [`Regex` type](#regex-type)
+        - [Capture type](#capture-type)
+            - [Basics](#basics)
+            - [Capturing group](#capturing-group)
+            - [Concatenation](#concatenation)
+            - [Named capturing group](#named-capturing-group)
+            - [Non-capturing group](#non-capturing-group)
+            - [Nested capturing group](#nested-capturing-group)
+            - [Quantification](#quantification)
+            - [Alternation](#alternation)
+    - [Effect on ABI stability](#effect-on-abi-stability)
+    - [Effect on API resilience](#effect-on-api-resilience)
+    - [Alternatives considered](#alternatives-considered)
+        - [Lazy collections instead of arrays of substrings](#lazy-collections-instead-of-arrays-of-substrings)
+        - [Homogeneous tuples for exact-count quantification](#homogeneous-tuples-for-exact-count-quantification)
+        - [`Never` as empty capture instead of `Void`](#never-as-empty-capture-instead-of-void)
+    - [Future directions](#future-directions)
+        - [Dynamic captures](#dynamic-captures)
+
+<!-- markdown-toc end -->
+
+## Introduction
+
+Capturing groups are a commonly used component of regular expressions as they
+allow the programmer to extract information from matched input. A capturing
+group collects multiple characters together as a single unit that can be
+[backreferenced](https://www.regular-expressions.info/backref.html) within the
+regular expression and accessed in the result of a successful match. For
+example, the following regular expression contains the capturing groups `(cd*)`
+and `(ef)`.
+
+```swift
+// Literal version.
+let regex = /ab(cd*)(ef)gh/
+// => `Regex<(Substring, Substring)>`
```

Comment:
I'm pretty sure we don't have to store `Substring` or any other views in the implementation. It can be all ranges until the point `.captures` is called to access the captures.